incubator/hnc/pkg/validators/hierarchical_namespace.go

```diff
@@ -0,0 +1,122 @@
+type hnsRequest struct {
+	hns *api.HierarchicalNamespace
+	op  v1beta1.Operation
+}
+
+// Handle implements the validation webhook.
+func (v *HierarchicalNamespace) Handle(ctx context.Context, req admission.Request) admission.Response {
+	log := v.Log.WithValues("Namespace", req.Namespace, "Name", req.Name)
+	// Early exit since the HNC SA can do whatever it wants.
+	if isHNCServiceAccount(&req.AdmissionRequest.UserInfo) {
+		log.V(1).Info("Allowed change by HNC SA")
+		return allow("HNC SA")
+	}
+
+	decoded, err := v.decodeRequest(log, req)
+	if err != nil {
+		v.Log.Error(err, "Couldn't decode request")
+		return deny(metav1.StatusReasonBadRequest, err.Error())
+	}
+
+	resp := v.handle(log, decoded)
+	log.V(1).Info("Handled", "allowed", resp.Allowed, "code", resp.Result.Code, "reason", resp.Result.Reason, "message", resp.Result.Message)
+	return resp
+}
+
+// handle implements the non-boilerplate logic of this validator, allowing it to be more easily unit
+// tested (ie without constructing a full admission.Request). It validates that the request is allowed
+// based on the current in-memory state of the forest.
+func (v *HierarchicalNamespace) handle(log logr.Logger, req *hnsRequest) admission.Response {
+	v.Forest.Lock()
+	defer v.Forest.Unlock()
+
+	pnm := req.hns.Namespace
+	cnm := req.hns.Name
+	cns := v.Forest.Get(cnm)
+
+	if req.op == v1beta1.Create {
+		if reconcilers.EX[pnm] {
+			msg := fmt.Sprintf("The namespace %s is not allowed to create subnamespaces. Please create hns in a different namespace.", pnm)
+			log.V(1).Info(msg)
+			return deny(metav1.StatusReasonForbidden, msg)
+		}
+
+		if cns.Exists() {
+			msg := fmt.Sprintf("The requested namespace %s already exists. Please use a different name.", cnm)
+			log.V(1).Info(msg)
+			return deny(metav1.StatusReasonConflict, msg)
+		}
+	}
+
+	if req.op == v1beta1.Delete {
+		if cns.Exists() && !cns.AllowsCascadingDelete() {
+			msg := fmt.Sprintf("The subnamespace %s doesn't allow cascading deletion. Please set allowCascadingDelete flag first.", cnm)
+			log.V(1).Info(msg)
+			return deny(metav1.StatusReasonForbidden, msg)
+		}
+	}
+
+	return allow("")
+}
+
+// decodeRequest gets the information we care about into a simple struct that's easy to both a) use
+// and b) factor out in unit tests.
+func (v *HierarchicalNamespace) decodeRequest(log logr.Logger, in admission.Request) (*hnsRequest, error) {
+	hns := &api.HierarchicalNamespace{}
+	var err error
+	// For DELETE request, use DecodeRaw() from req.OldObject, since Decode() only uses req.Object,
+	// which will be empty for a DELETE request.
+	if in.Operation == v1beta1.Delete {
+		log.V(1).Info("Decoding a delete request.")
+		err = v.decoder.DecodeRaw(in.OldObject, hns)
+	} else {
+		err = v.decoder.Decode(in, hns)
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	return &hnsRequest{
+		hns: hns,
+		op:  in.Operation,
+	}, nil
+}
+
+func (v *HierarchicalNamespace) InjectDecoder(d *admission.Decoder) error {
```

Comment:
Is this function used in this PR? I'd expect to see a test for something like this, but don't see anything associated with it -- either that or as a part of an interface definition. Am I missing some understanding here?

Reply:
You mean `InjectDecoder`? It's part of the boilerplate code that's called by `controller-runtime`, I'm pretty sure nothing will compile if it's not there.

Reply:
The documentation of the interface DecoderInjector is here: https://godoc.org/sigs.k8s.io/controller-runtime/pkg/webhook/admission#DecoderInjector.

Reply:
I tried just deleting `InjectDecoder` and `InjectClient` methods out of .pkg/validators/hierarchy.go as a test and the binary still compiles and deploys. I'm not one to argue with the compiler. As an experiment can you see what'd happen if we took this out and see if there's any side effects? I don't want to block the review or anything -- you're just likely closest to trying this out!

Reply:
Hmm. Are you sure the webhook is running correctly after this? Without `InjectDecoder`, I'm not sure how the `decoder` field is set, which means it should cause a nil pointer exception the first time we try to decode something. Perhaps it just looks fine because you never created a HierarchicalNamespace object, but as soon as you do, HNC will crash?

I would have _thought_ that `InjectDecoder` was checked at compile time, but perhaps it's only checked at runtime with a [type switch](https://tour.golang.org/methods/16).

Reply:
Thanks for bringing this up! I am also curious so I tested with removing the `InjectDecoder` in namespace webhook and `make deploy`. I got errors when trying to delete a namespace. The errors was the same as when I had dereference of an nil pointer similar to #603 .

Reply:
Why didn't you need the decoder when trying to _create_ a namespace?

Reply:
I was just testing if it would work if the `InjectDecoder` is removed from the code. And it turned out it will fail, so we need it.

Reply:
👍